File: cli/backend/meteo/meteo_current.py
"""Alternative Weather Backend"""
import time
from datetime import datetime

from weather_core import WeatherFile
from weather_core.backend import WindData, WeatherCondition, WeatherData


class MeteoCurrent(WeatherData):
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, forecast_json, aqi_json, metric):
        self.now = forecast_json["hourly"]["time"].index(
            forecast_json["current_weather"]["time"]
        )
        wind = WindData(
            forecast_json["current_weather"]["windspeed"],
            int(forecast_json["current_weather"]["winddirection"]),
        )
        cloud_cover = forecast_json["hourly"]["cloudcover"][self.now]
        date = datetime.fromisoformat(forecast_json["current_weather"]["time"])
        self.time = int(time.mktime(date.timetuple()))
        self.temperature = round(forecast_json["current_weather"]["temperature"], 2)
        self.min_temp = forecast_json["daily"]["temperature_2m_min"][0]
        self.max_temp = forecast_json["daily"]["temperature_2m_max"][0]
        self.wind = wind
        self.dewpoint = round(forecast_json["hourly"]["dewpoint_2m"][self.now], 2)
        self.feels_like = round(
            forecast_json["hourly"]["apparent_temperature"][self.now], 2
        )
        self.aqi = aqi_json["hourly"]["european_aqi"][self.now] // 20
        self.cloud_cover = cloud_cover
        conditions = []
        weather_codes = WeatherFile.weather_codes().data
        if cloud_cover == 0:
            conditions.append(WeatherCondition(800, weather_codes))
        elif cloud_cover < 25:
            conditions.append(WeatherCondition(801, weather_codes))
        elif cloud_cover < 50:
            conditions.append(WeatherCondition(802, weather_codes))
        elif cloud_cover < 85:
            conditions.append(WeatherCondition(803, weather_codes))
        else:
            conditions.append(WeatherCondition(804, weather_codes))
        if forecast_json["hourly"]["rain"][self.now] != 0:
            if (
                0 < forecast_json["hourly"]["rain"][self.now] < 0.098 and not metric
            ) or (0 < forecast_json["hourly"]["rain"][self.now] < 2.5 and metric):
                conditions.append(WeatherCondition(500, weather_codes))
            elif (forecast_json["hourly"]["rain"][self.now] < 0.39 and not metric) or (
                forecast_json["hourly"]["rain"][self.now] < 10 and metric
            ):
                conditions.append(WeatherCondition(501, weather_codes))
            elif (forecast_json["hourly"]["rain"][self.now] < 2 and not metric) or (
                forecast_json["hourly"]["rain"][self.now] < 50 and metric
            ):
                conditions.append(WeatherCondition(502, weather_codes))
            elif forecast_json["hourly"]["rain"][self.now] != 0:
                conditions.append(WeatherCondition(503, weather_codes))
        if forecast_json["hourly"]["snowfall"][self.now] != 0:
            conditions.append(WeatherCondition(601, weather_codes))
        self.conditions = conditions
        self.condition_ids = self.get_condition_ids()
        self.condition_sentence = self.get_conditions_sentence()

File: src/backend/meteo/meteo_current.rs
use crate::backend::meteo::meteo_json::{MeteoAirQualityJson, MeteoForecastJson};
use crate::backend::weather_condition::WeatherCondition;
use crate::backend::weather_data::{get_conditions_sentence, WeatherDataRS};
use crate::backend::wind_data::WindData;
use crate::local::weather_file::WeatherFile;
use crate::now;

pub struct MeteoCurrent {
    data: MeteoForecastJson,
    aqi: MeteoAirQualityJson,
    pub index: usize,
    metric: bool,
}

impl MeteoCurrent {
    pub fn new(data: MeteoForecastJson, aqi: MeteoAirQualityJson, metric: bool) -> Self {
        let index = data.hourly.time.iter().position(|r| *r == data.current_weather.time).unwrap();
        MeteoCurrent {
            data,
            aqi,
            metric,
            index
        }
    }
}

impl WeatherDataRS for MeteoCurrent {
    fn get_time(&self) -> i128 {
        now() as i128
    }

    fn get_temperature(&self) -> f32 {
        self.data.current_weather.temperature
    }

    fn get_min_temp(&self) -> f32 {
        self.data.daily.temperature_2m_min[0]
    }

    fn get_max_temp(&self) -> f32 {
        self.data.daily.temperature_2m_max[0]
    }

    fn get_wind(&self) -> WindData {
        WindData {
            speed: self.data.current_weather.windspeed as f64,
            heading: self.data.current_weather.winddirection as i16,
        }
    }

    fn get_raw_data(&self) -> String {
        serde_json::to_string_pretty(&self.data).expect("dump to string failed")
    }

    fn get_dewpoint(&self) -> f32 {
        self.data.hourly.dewpoint_2m[self.index]
    }

    fn get_feels_like(&self) -> f32 {
        self.data.hourly.apparent_temperature[self.index]
    }

    fn get_aqi(&self) -> u8 {
        self.aqi.hourly.european_aqi[self.index]
    }

    fn get_cloud_cover(&self) -> u8 {
        self.data.hourly.cloudcover[self.index]
    }

    fn get_conditions(&self) -> Vec<WeatherCondition> {
        let weather_file = WeatherFile::weather_codes();
        let mut conditions: Vec<WeatherCondition> = Vec::new();
        let cloud_cover = self.get_cloud_cover();
        if cloud_cover == 0 {
            conditions.push(WeatherCondition::new(800, &weather_file.data));
        }
        else if cloud_cover < 25
        {
            conditions.push(WeatherCondition::new(801, &weather_file.data));
        }
        else if cloud_cover < 50 {
            conditions.push(WeatherCondition::new(802, &weather_file.data));
        }
        else if cloud_cover < 85 {
            conditions.push(WeatherCondition::new(803, &weather_file.data));
        }
        else {
            conditions.push(WeatherCondition::new(804, &weather_file.data));
        }
        if self.data.hourly.rain[self.index] != 0.0 {
            let rain = self.data.hourly.rain[self.index];
            let metric = self.metric;
            if (0.0 < rain && rain < 0.098 && !metric
            ) || (0.0 < rain && rain  < 2.5 && metric) {
                conditions.push(WeatherCondition::new(500, &weather_file.data));
            } else if (rain < 0.39
            && !metric) || (
                rain < 10.0 && metric
            )
            {
                conditions.push(WeatherCondition::new(501, &weather_file.data));
            }
            else if (rain < 2.0 && !metric) || (
                rain < 50.0 && metric
            )
            {
                conditions.push(WeatherCondition::new(502, &weather_file.data));
            }
            else if rain != 0.0 {
                conditions.push(WeatherCondition::new(503, &weather_file.data));
            }
        }
        if self.data.hourly.snowfall[self.index] != 0.0 {
            conditions.push(WeatherCondition::new(601, &weather_file.data));
        }
        return conditions;
    }

    fn get_condition_sentence(&self) -> String {
        get_conditions_sentence(self.get_conditions())
    }
}